pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1340,111 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
+		}
+		return pullSecret, nil
+	case globalPullSecret != "":
+		return globalPullSecret, nil
+	case localPullSecret != "":
+		return localPullSecret, nil
+	default:
+		errMsg := "clusterdeployment must specify pull secret since hiveconfig does not specify a global pull secret"
+		cdLog.Error(errMsg)
+		return "", fmt.Errorf("%s", errMsg)
+	}
+}
+
+// updatePullSecretInfo adds pull secret information in cluster deployment and cluster deployment status
+func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+	var secretName string
+
+	// Get existing pull secret
+	isPullSecretObjExists := true
+	existingPullSecretObj := &corev1.Secret{}
+	if cd.Status.PullSecret.Name != "" {
+		secretName = cd.Status.PullSecret.Name
+		err := r.Get(context.TODO(), types.NamespacedName{Name: secretName, Namespace: cd.Namespace}, existingPullSecretObj)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				isPullSecretObjExists = false
+			} else {
+				return fmt.Errorf(fmt.Sprintf("Error getting pull secret from cluster deployment : %v", err))
+			}
+		}
+	} else {
+		isPullSecretObjExists = false
+	}
+
+	if isPullSecretObjExists {
+		existingPullSecret, ok := existingPullSecretObj.Data[corev1.DockerConfigJsonKey]
+		if !ok {
+			return fmt.Errorf("Pull secret %s did not contain key %s", secretName, corev1.DockerConfigJsonKey)
+		}
+		hashOfexistingSecret, err := controllerutils.GetHash(existingPullSecret)
+		if err != nil {
+			return fmt.Errorf(fmt.Sprintf("Error creating checksum of %s : %v", secretName, err))
+		}
+		hashOfNewSecret, err := controllerutils.GetHash([]byte(pullSecret))
+		if err != nil {
+			return fmt.Errorf(fmt.Sprintf("Error creating checksum of new pull secret : %v", err))
+		}
+
+		// Check if the pull secrets checksums  are  different
+		if hashOfexistingSecret == hashOfNewSecret {
+			cdLog.Debug("No new pull secret found")
+			return nil
+		}
+	}
+
+	// delete the existing pull secret object
+	if isPullSecretObjExists {
+		err := r.Delete(context.TODO(), existingPullSecretObj, client.PropagationPolicy(metav1.DeletePropagationBackground))
+		if err != nil {
+			return fmt.Errorf(fmt.Sprintf("Error deleting the old existing pull secret %s: %v", secretName, err))
+		}
+	}
+
+	// create a new pull secret object and update the cluster deployment status
+	newSecretName := apihelpers.GetResourceName(cd.Name, "merged-pullsecret-"+utilrand.String(8))
+	newPullSecretObj := generatePullSecretObj(pullSecret, newSecretName, cd)
+	err := r.Create(context.TODO(), newPullSecretObj)
+	if err != nil {
+		return fmt.Errorf(fmt.Sprintf("error creating new pull secret object: %v", err))
+	}
+	cdLog.Debug("Created the new pull secret object in cluster deployment successfully")
+
+	// Update new pullsecret information in cluster deployment status
+	cd.Status.PullSecret = corev1.LocalObjectReference{Name: newSecretName}
+	err = r.statusUpdate(cd, cdLog)
+	defer func(err error) {
```

Comment:
How does it makes any difference? As I am calling defer after `err = r.statusUpdate(cd, cdLog)` , so what ever value err would have that time will be taken by defer. example https://play.golang.org/p/76R-ar5ia-D

Reply:
From your example, the point is that you want defer to print 15, not 5.

Reply:
You want to use the value of err when the function finishes not when the defer is establishes. Imagine that there was some other task that needed to be done after setting the clusterdeployment status. You want the secret to be deleted if either of those tasks failed.

Reply:
Really, the preferred way of doing this would be to name the return parameter of the `updatePullSecretInfo` function and look at that value rather than the local value.

Reply:
https://play.golang.org/p/Xn7EJZhqeAI

Reply:
The goal of the defer is only to delete the new pull secret object created if the status update fails. Why would we want to run `err := r.Delete(context.TODO(), newPullSecretObj)` for any errors?

Reply:
The goal of the defer is to ensure that the entire operation is atomic. Either everything succeeds or nothing happened. So any future error in the function should revert the creation of the secret. As of now, the only future error would be from updating the clusterdeployment status, but the code should still be written in such a way that adding some other possibly failing task in the future will not affect whether the creation of the secret gets properly reverted.

Reply:
That makes sense. To make it an atomic operation I need to move the deletion of old pull secret object after status update. Made changes to the PR. Let me know if you disagree.